server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -652,4 +652,15 @@ public long getMinUnreferencedSeqNoInSegments(long minUnrefCheckpointInLastCommi
     int availablePermits() {
         return syncPermit.availablePermits();
     }
+
+    /**
+     * Checks whether or not the shard should be refreshed.
+     * This checks if number of translog files breaches the threshold count determined by
+     * {@code cluster.remote_store.translog.max_readers} setting
+     * @return {@code true} if the shard should be refreshed
+     */
+    @Override
+    public boolean shouldRefreshShard() {
+        return readers.size() >= translogTransferManager.getMaxRemoteTranslogReadersSettings();
```

Comment:
There is a setting for keeping extra translog `index.remote_store.translog.keep_extra_gen`. We have readers to delete lower than this, the system can become unpredictable. Lets add validations and checks for the same.

Reply:
That is only applicable for remote translog trimming, hence won't interfere.